xds/internal/xdsclient/transport/lrs/lrs_stream.go

```diff
@@ -0,0 +1,326 @@
+	node.ClientFeatures = append(node.ClientFeatures, "envoy.lrs.supports_send_all_clusters")
+
+	runLoadReportStream := func() error {
+		// streamCtx is created and canceled in case we terminate the stream
+		// early for any reason, to avoid gRPC-Go leaking the RPC's monitoring
+		// goroutine.
+		streamCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream, err := lrs.transport.CreateStreamingCall(ctx, "/envoy.service.load_stats.v3.LoadReportingService/StreamLoadStats")
+		if err != nil {
+			lrs.logger.Warningf("Creating new LRS stream failed: %v", err)
+			return nil
+		}
+		if lrs.logger.V(2) {
+			lrs.logger.Infof("LRS stream created")
+		}
+
+		if err := lrs.sendFirstLoadStatsRequest(stream, node); err != nil {
+			lrs.logger.Warningf("Sending first LRS request failed: %v", err)
+			return nil
+		}
+
+		clusters, interval, err := lrs.recvFirstLoadStatsResponse(stream)
+		if err != nil {
+			lrs.logger.Warningf("Reading from LRS stream failed: %v", err)
+			return nil
+		}
+
+		// We reset backoff state when we successfully receive at least one
+		// message from the server.
+		lrs.sendLoads(streamCtx, stream, clusters, interval)
+		return backoff.ErrResetBackoff
+	}
+	backoff.RunF(ctx, runLoadReportStream, lrs.backoff)
+}
+
+func (lrs *Stream) sendLoads(ctx context.Context, stream transport.StreamingCall, clusterNames []string, interval time.Duration) {
+	tick := time.NewTicker(interval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+		case <-ctx.Done():
+			return
+		}
+		if err := lrs.sendLoadStatsRequest(stream, lrs.lrsStore.Stats(clusterNames)); err != nil {
+			lrs.logger.Warningf("Writing to LRS stream failed: %v", err)
+			return
+		}
+	}
+}
+
+func (lrs *Stream) sendFirstLoadStatsRequest(stream transport.StreamingCall, node *v3corepb.Node) error {
+	req := &v3lrspb.LoadStatsRequest{Node: node}
+	if lrs.logger.V(perRPCVerbosityLevel) {
+		lrs.logger.Infof("Sending initial LoadStatsRequest: %s", pretty.ToJSON(req))
+	}
+	err := stream.Send(req)
+	if err == io.EOF {
+		return getStreamError(stream)
+	}
+	return err
+}
+
+func (lrs *Stream) recvFirstLoadStatsResponse(stream transport.StreamingCall) ([]string, time.Duration, error) {
+	r, err := stream.Recv()
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to receive first LoadStatsResponse: %v", err)
+	}
+	resp, ok := r.(*v3lrspb.LoadStatsResponse)
+	if !ok {
+		lrs.logger.Infof("Message received on ADS stream of unexpected type: %T", r)
+		return nil, time.Duration(0), fmt.Errorf("unexpected message type %T", r)
+	}
+	if lrs.logger.V(perRPCVerbosityLevel) {
+		lrs.logger.Infof("Received first LoadStatsResponse: %s", pretty.ToJSON(resp))
+	}
+
+	rInterval := resp.GetLoadReportingInterval()
+	if rInterval.CheckValid() != nil {
+		return nil, 0, fmt.Errorf("invalid load_reporting_interval: %v", err)
+	}
+	interval := rInterval.AsDuration()
+
+	if resp.ReportEndpointGranularity {
+		// TODO(easwars): Support per endpoint loads.
```

Comment:
We should move all the watches to resolver before this probably to prevent config tearing, which would affect the correctness of LRS too I think.

Reply:
I don't quite understand. We will handle config tears when we come around to handling it. It is a non-trivial amount of work. But blocking a refactor which mostly tries to retain existing functionality which is required to unblock other things doesn't seem like the correct option to me.

Reply:
Isn't this TODO of per endpoint loads technically broken if we have config tears still?

Reply:
This is a TODO that is present in the current code. And we don't handle config tears yet.

I still don't understand where you are coming from though. Config tears can lead to a bunch of problems and if LRS is one of them, so be it, because those problems exist today. 

But if you are saying that this TODO somehow cannot be implemented without adding support to handle config tears, I'd like to hear about that, and maybe we can clarify the TODO.

Reply:
"Config tears can lead to a bunch of problems and if LRS is one of them, so be it, because those problems exist today." - ok sounds good, I meant it like what's the point of doing it if it's incorrect with respect to EDS. So I guess this TODO is fine, and something we eventually want to do orthogonal to the config tear fix.